### PR TITLE
[inductor] Optimize masks based on block size

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #116360
* #116079

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/codegen/triton.py [hunk 1]
```diff
@@ -1023,7 +1023,22 @@ def codegen_header(self, code, no_x_dim=False):
                     f"{self.name} = {line}",
                 ]
             )
-        code.writeline(f"{x}mask = {self.name} < {x}numel")
+
+        max_block = config.triton.max_block[x.upper()]
+        no_mask_threshold = int(sympy.gcd(self.numel, max_block))
```
Comment (lezcano) [inline torch/_inductor/codegen/triton.py:1028]: better to see look at `self.numel % max_bock == 0`.

**Review by lezcano — COMMENTED** [review]
We already implemented something like this at https://github.com/pytorch/pytorch/pull/91254/. It would be best to merge both approaches, i.e., fully avoiding generating the code when the mask is divisible. If the masks are not currently used when necessary given the other PR, the mask creation code will likely be DCE'd by triton, so this will most likely be a cosmetic change (which is itself nice to improve readability, but yeah).

File: torch/_inductor/codegen/triton.py (hunk 1 above)
Comment (jansel) [reply torch/_inductor/codegen/triton.py:1028]: That would give me a boolean, not a threshold that can be checked after the autotuner picks the block size.

**Comment by jansel** [comment]
> We already implemented something like this at #91254. It would be best to merge both approaches, i.e., fully avoiding generating the code when the mask is divisible. If the masks are not currently used when necessary given the other PR, the mask creation code will likely be DCE'd by triton, so this will most likely be a cosmetic change (which is itself nice to improve readability, but yeah).

The issue is with the autotuner sometimes a mask can be removed, and sometimes it can't, based on which config is chosen.

**Review by lezcano — APPROVED** [review]

**Pull request closed without merging**